### PR TITLE
[UAR-763] statement validation middleware run on answer of 2nd statement

### PR DESCRIPTION
**Pull request by corybudd-ch**
### JIRA link
https://companieshouse.atlassian.net/browse/UAR-763

### Change description
Introduced middleware that will validate the statements the user has made in their journey, and the middleware will (eventually) contain a set of rules that checks their statements are correct in relation to what they have answered/added in the BO/MO journey.
This change is for the 1st rule check, checking that at least 1 active BO exists if they have chosen they can identify all or some BOs in the 1st statement page `Has the overseas entity identified any registrable beneficial owners?`.

I've set the default behaviour after the 2nd statement page to redirect to the `update-statement-validation-errors` page, then the middleware runs on this route, so if statement validation fails it allows the middleware to manipulate the req, by adding req['statementErrorList'] and calling next(), which will continue onto the `update-statement-validation-errors` page which can then access the `statementErrorList` and display the errors.
Or if the user's statements are correct, the middleware will redirect the user away from `update-statement-validation-errors` page and onto their next page.

Screenshot 2023-08-17 at 09 17 05

### Work checklist

- [ ] Tests added where applicable
- [ ] UI changes meet accessibility criteria

### Merge instructions

We are committed to keeping commit history clean, consistent and linear. To achieve this, this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

File: src/utils/update/psc.to.beneficial.owner.type.mapper.ts [hunk 1]
```diff
@@ -21,6 +21,7 @@ export const mapPscToBeneficialOwnerTypeIndividual = (psc: CompanyPersonWithSign
     usual_residential_address: undefined,
     service_address: service_address,
     start_date: mapInputDate(psc.notifiedOn),
+    ceased_date: psc.ceasedOn ? mapInputDate(psc.ceasedOn) : undefined,
```
[the remark this answers is not in the document]
Comment (corybudd-ch) [reply src/utils/update/psc.to.beneficial.owner.type.mapper.ts:24]: Yeah this was for my own purposes to test `checkActiveBOExists`.

So ... I was confused about this. Why are we not mapping ceasedOn date? What's the point of this statement validation if we're going to consider every BO as active? And if we mark BOs as ceased in an update journey, then on further updates when we retrieve the data we'll consider them active?

[the remark this answers is not in the document]
Comment (corybudd-ch) [reply src/utils/update/psc.to.beneficial.owner.type.mapper.ts:24]: I'm with you now, so we ignore retrieving BOs that have a ceasedOn date whenever we are doing the mapping on the Web side .. hmm need to think how that impacts this

[the remark this answers is not in the document]
Comment (AndreKramer7654) [reply src/utils/update/psc.to.beneficial.owner.type.mapper.ts:24]: Assuming ceased dates are in the past so no need to review any that are ceased. There's no harm in mapping the ceased but wanted to point out that we currently filter those out anyway.

File: src/controllers/update/update.statement.validation.errors.controller.ts [hunk 2]
```diff
@@ -36,22 +38,23 @@ export const post = (req: Request, res: Response, next: NextFunction) => {
...
 
     if (!errors.isEmpty()) {
-      return renderPage(res, appData, inNoChangeJourney, formatValidationError(errors.array()));
+      return renderPage(res, appData, inNoChangeJourney, req.body['statementErrorList[]'], formatValidationError(errors.array()));
     }
 
     return req.body[StatementResolutionKey] === StatementResolutionType.CHANGE_INFORMATION
       ? res.redirect(getChangeInformationRedirectUrl(inNoChangeJourney))
-      : res.redirect(getBOStatementsUrl(inNoChangeJourney));
+      : res.redirect(getFirstBOStatementsUrl(inNoChangeJourney));
```
Comment (stephanie-ch) [inline src/controllers/update/update.statement.validation.errors.controller.ts:46]: how about identifiedBOStatement and ceasedBOStatement?

File: src/middleware/statement.validation.middleware.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  ) {
+    req['statementErrorList'] = errorList;
+    return next();
+  }
+
+  const redirectUrl = appData.update?.no_change ? UPDATE_REVIEW_STATEMENT_URL : UPDATE_CHECK_YOUR_ANSWERS_URL;
+  return res.redirect(redirectUrl);
+};
+
+const validateIdentifiedBOsStatement = (appData: ApplicationData, errorList: string[]): boolean => {
```
Comment (stephanie-ch) [inline src/middleware/statement.validation.middleware.ts:30]: Do we want to move this to somewhere outside of the middleware? For reuse on the nav ticket?

Comment (stephanie-ch) [reply src/middleware/statement.validation.middleware.ts:30]: Discussed, will do in later ticket.

File: src/controllers/update/update.statement.validation.errors.controller.ts (hunk 2 above)
Comment (corybudd-ch) [reply src/controllers/update/update.statement.validation.errors.controller.ts]: updated - thank you!

File: views/update/update-statement-validation-errors.html [hunk 4]
```diff
@@ -93,6 +93,9 @@ <h1 class="govuk-heading-xl">{{ title }}</h1>
       }) }}
 
       {% include "includes/continue-button.html" %}
+      {% for statementError in statementErrorList %}
```
Comment (stephanie-ch) [inline views/update/update-statement-validation-errors.html:96]: Should we add a comment here, about why? Or is it clear enough?

**Review by cdrainCH — APPROVED** [review]

**Review by joshuahigginsonCH — APPROVED** [review]

File: views/update/update-statement-validation-errors.html (hunk 4 above)
Comment (corybudd-ch) [reply views/update/update-statement-validation-errors.html:98]: done

**Pull request merged**